src/goal.rs

```diff
@@ -904,6 +963,60 @@ impl<'a> Goal<'a> {
         ProofTerm::CaseSplit(var_str, instantiated_cons_and_goals),
       );
     }
+
+    CaseSplitStatus::Success
+  }
+
+  fn sexp_is_constructor(&self, sexp: &Sexp) -> bool {
+    match sexp {
+      Sexp::String(s) => is_constructor(s),
+      Sexp::List(v) => is_constructor(&v[0].string().unwrap()),
+      _ => false,
+    }
+  }
+
+  fn gen_fresh_vars(&self, sexp: Sexp, mut idx: i32) -> Sexp {
+    let qm = "?".to_string();
+    match sexp {
+      Sexp::String(s) if s == qm => Sexp::String(format!("?block_{}", idx)),
+      Sexp::List(v) => Sexp::List(
+        v.iter()
+          .map(|x| {
+            idx = idx + 1;
+            self.gen_fresh_vars(x.clone(), idx + 1)
+          })
+          .collect(),
+      ),
+      _ => sexp,
+    }
+  }
+
+  fn analyze_sexp_for_blocking_vars(&self, sexp: &Sexp) -> Vec<Sexp> {
```

Comment:
So this generates variants of a single searcher where we replace every part of it with a variable, e.g.

```
(add (S ?x) (S ?y))
```
with
```
(add ?fresh (S ?y))
(add (S ?x) ?fresh)
(add ?fresh1 ?fresh2)
```